src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -25,16 +35,121 @@
 
   /*
    * Called when a client submits a GET request to the /data URL
-   * Displays all recorded user comments on page
    */
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {}
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // PROTO Data structure:
+    // Parse the contents  of graph.txt into a proto Graph object, and extract information
+    // from the proto object into a map. This is used to store the proto input and isn't updated
+    // with mutations.
+    Graph protoGraph =
+        Graph.parseFrom(getServletContext().getResourceAsStream("/WEB-INF/graph.txt"));
+    Map<String, Node> protoNodesMap = protoGraph.getNodesMapMap();
+
+    // GRAPH Data structures:
+    // Create an undirected graph data structure to store the information, and
+    // map each node name in the graph to the GraphNode objects. This is the graph & map
+    // we update with mutations
+    MutableGraph<GraphNode> graph = GraphBuilder.directed().build();
+    HashMap<String, GraphNode> graphNodesMap = new HashMap<>();
+
+    // Create a graph (useable form) from proto data structures
+    for (String nodeName : protoNodesMap.keySet()) {
+
+      Node thisNode = protoNodesMap.get(nodeName);
+
+      // Convert thisNode into a graph node that may store additional information
+      GraphNode graphNode = protoNodeToGraphNode(thisNode);
+
+      // Update graph data structures to include the node
+      graph.addNode(graphNode);
+      graphNodesMap.put(nodeName, graphNode);
+
+      // Add dependency edges to the graph
+      for (String child : thisNode.getChildrenList()) {
+        graph.putEdge(graphNode, protoNodeToGraphNode(protoNodesMap.get(child)));
```

Comment:
We were relying on [this](https://guava.dev/releases/23.0/api/docs/com/google/common/graph/MutableGraph.html#putEdge-N-N-) implementation dependent behavior that if the node exists an edge is added but if it isn't both the node and an edge are added. I added an extra check for this case, but I'm not sure how we would end up with duplicate nodes?

Reply:
I think I see what you mean, but the graph needs nodes to be unique so it is probably using the equals method of graph nodes to prevent duplicates and these two objects would be equal?

Reply:
Ah, I see. We can make sure that this is the case with some unit tests ;)

Reply:
Yup I did this